TeensyDmx.h

```diff
@@ -34,6 +34,62 @@
 #define UART_C3_FEIE    (uint8_t)0x02   // Framing Error Interrupt Enable
 #endif
 
+// ----- macros -----
+
+// 16-bit and 32-bit integers in the RDM protocol are transmitted highbyte - lowbyte.
+// but the ATMEGA processors store them in highbyte - lowbyte order.
+// Use SWAPINT to swap the 2 bytes of an 16-bit int to match the byte order on the DMX Protocol.
+// avoid using this macro on variables but use it on the constant definitions.
+#define SWAPINT(i) (((i&0x00FF)<<8) | ((i&0xFF00)>>8))
+// Use SWAPINT32 to swap the 4 bytes of a 32-bit int to match the byte order on the DMX Protocol.
+#define SWAPINT32(i) ((i&0x000000ff)<<24) | ((i&0x0000ff00)<<8) | ((i&0x00ff0000)>>8) | ((i&0xff000000)>>24)
+
+// read a 16 bit number from a data buffer location
+#define READINT(p) ((p[0]<<8) | (p[1]))
+
+enum { DMX_BUFFER_SIZE = 512 };
+enum { RDM_BUFFER_SIZE = (24+231) }; //TODO(Peter): Check me
+
+struct RDMDATA
+{
+  byte     StartCode;    // Start Code 0xCC for RDM
+  byte     SubStartCode; // Start Code 0x01 for RDM
+  byte     Length;       // packet length
+  byte     DestID[6];
+  byte     SourceID[6];
+
+  byte     _TransNo;     // transaction number, not checked
+  byte     ResponseType;    // ResponseType or PortID
+  byte     MessageCount;     // number of queued messages
+  uint16_t SubDev;      // sub device number (root = 0)
+  byte     CmdClass;     // command class
+  uint16_t Parameter;	   // parameter ID
+  byte     DataLength;   // parameter data length in bytes
+  byte     Data[231];   // data byte field
+} __attribute__((__packed__)); // struct RDMDATA
+
+// the special discovery response message
+struct DISCOVERYMSG
+{
+  byte headerFE[7];
+  byte headerAA;
+  byte maskedDevID[12];
+  byte checksum[4];
+} __attribute__((__packed__)); // struct DISCOVERYMSG
+
+// The buffer for RDM packets being received and sent.
+// this structure is needed to seperate RDM data from DMX data.
+union RDMMSG {
+  // the most common RDM packet layout for commands
+  struct RDMDATA packet;
+
+  // the layout of the RDM packet when returning a discovery message
+  struct DISCOVERYMSG discovery;
+
+  // the byte array used while receiving and sending.
+  byte buffer[RDM_BUFFER_SIZE];
```

Comment:
Can the DMX buffer not be used to save space?

Reply:
Yes possibly looking at it. It was something I did as part of the troubleshooting, when stuff wasn't working. Currently the code works as follows from what I can see:

* DMX Tx, set active, write out active
** RDM could Tx/Rx via inactive buffer
* DMX Rx, get inactive, read into active, swap on complete frame
** RDM could Rx/Tx via active buffer (only swap on complete DMX frame)

So it looks like it's technically possibly, but will need to swap which buffer it's using for RDM depending on what the device is doing (DMX Tx or Rx).

I wonder if actually you should be double buffering the DMX Tx and swapping buffers, to give someone a chance of writing a complete frame, rather than part of a frame, then it transmitting it, then the rest of the frame? This would also mean getBuffer would work in both scenarios, so you could set individual channels in Tx, but be able to use getBuffer to find the overall state of what you're sending. Otherwise the user will just have to have their own private buffer to keep track of that.